flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -223,16 +214,25 @@
     private boolean isTriggering = false;
 
     private final CheckpointRequestDecider requestDecider;
+
+    private final CheckpointBriefCalculator checkpointBriefCalculator;
+
+    private final ExecutionAttemptMappingProvider attemptMappingProvider;
+
     private final LinkedHashMap<ExecutionAttemptID, ExecutionVertex> cachedTasksById;
 
+    /**
+     * Temporary flag to allow checkpoints after tasks finished. This is disabled for regular jobs
+     * to keep the current behavior but we want to allow it in tests. This should be removed once
+     * all parts of the stack support checkpoints after some tasks finished.
+     */
+    private boolean allowCheckpointsAfterTasksFinished;
```

Comment:
It is currently used in `calculateCheckpointBrief()` to disable checkpoints after tasks finished. Previously I'm thinking to enable this flag in tests, but now I realized it would makes more sense to keep the tests and the normal process consistent. I think it reveals some tests are missed for our scenarios and I'll complement the tests.

Reply:
Yes, this is not a feature toggle and would be removed in the final PR. 

The flag (and also the flag in the TM side) is required since if not all of this PRs are get in some behaviors would be not right. We would remove all these flags in the last PR, which enables the whole functionality atomically.